optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,329 @@
+            "by using `{self._independent_sampler.__class__.__name__}` instead of `QMCSampler` "
+            "(optimization performance may be degraded). "
+            "`QMCSampler` does not support dynamic search space or `CategoricalDistribution`. "
+            "You can suppress this warning by setting `warn_independent_sampling` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this independent sampling is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = trans.bounds[:, 0] + sample * (trans.bounds[:, 1] - trans.bounds[:, 0])
+        return trans.untransform(sample[0, :])
+
+    def after_trial(
+        self,
+        study: "optuna.Study",
+        trial: "optuna.trial.FrozenTrial",
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        self._independent_sampler.after_trial(study, trial, state, values)
+
+    def _sample_qmc(self, study: Study, search_space: Dict[str, BaseDistribution]) -> np.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, search_space)
+        d = len(search_space)
+
+        if self._qmc_type == "halton":
+            qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+        elif self._qmc_type == "sobol":
+            qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+        else:
+            raise ValueError("Invalid `qmc_type`")
+
+        forward_size = sample_id  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
+        sample = qmc_engine.random(1)
+
+        return sample
+
+    def _find_sample_id(self, study: Study, search_space: Dict[str, BaseDistribution]) -> int:
+
+        qmc_id = ""
+        qmc_id += self._qmc_type
+        # Sobol/Halton sequences without scrambling do not use seed.
+        if self._scramble:
+            qmc_id += f" (scramble=True, seed={self._seed})"
+        else:
+            qmc_id += " (scramble=False)"
+        key_qmc_id = qmc_id + "'s last sample id"
```

Comment:
Here, I meant that we can use some constant string such as `qmc_sampler:sample_id`
e.g. https://github.com/optuna/optuna/blob/master/optuna/integration/botorch.py#L471-L473

Why do we need to create a different key for each `scramble` and `seed` pair?

Reply:
Ok. I understand. But, if we really want to warn it to users, we may have better way to share the information. We do not need to add it in this PR, but we can update it in the future.